### PR TITLE
wgpu 0.17

### PR DESCRIPTION
**Pull request by Elabajaba**
~~Currently blocked on an upstream bug that causes crashes when minimizing/resizing on dx12 https://github.com/gfx-rs/wgpu/issues/3967~~ wgpu 0.17.1 is out which fixes it

# Objective

Keep wgpu up to date.

## Solution

Update wgpu and naga_oil.

Currently this depends on an unreleased (and unmerged) branch of naga_oil, and hasn't been properly tested yet.

The wgpu side of this seems to have been an extremely trivial upgrade (all the upgrade work seems to be in naga_oil). This also lets us remove the workarounds for pack/unpack4x8unorm in the SSAO shaders.

Lets us close the dx12 part of https://github.com/bevyengine/bevy/issues/8888

related: https://github.com/bevyengine/bevy/issues/9304

---

## Changelog

Update to wgpu 0.17 and naga_oil 0.9

### CODE REVIEW

**Comment by Elabajaba** [comment]
Looks like wasm basically needs a rewrite of the renderer...

**Comment by Elabajaba** [comment]
I'm not sure if we should add a panic somewhere in the renderer in case someone compiles for wasm with `-Ctarget-feature=+atomics` to warn users that it isn't currently supported, or just leave it for now until bevy supports wasm threading.

File: crates/bevy_render/Cargo.toml
Comment (JMS55) [inline crates/bevy_render/Cargo.toml:80]: Small nit: Can we move naga_oil up to underneath naga?

File: crates/bevy_render/Cargo.toml
Comment (JMS55) [inline crates/bevy_render/Cargo.toml:62]:
```suggestion
wgpu = { version = "0.17.0", features = ["naga", "fragile-send-sync-non-atomic-wasm"] }
```

**Review by fintelia — APPROVED** [review]

**Review by JMS55 — APPROVED** [review]

**Comment by Elabajaba** [comment]
Looks like minimising is broken on dx12 for w/e reason, just panics immediately when I minimise a window.

**Comment by Elabajaba** [comment]
https://github.com/gfx-rs/wgpu/pull/4106 is the upstream PR that fixes the crash, once that lands and there's a patch release this should be good?

**Comment by JMS55** [comment]
@Elabajaba status update on this PR?

**Comment by Elabajaba** [comment]
Still waiting on a patch release for wgpu.

**Comment by mockersf** [comment]
wgpu 0.17.1 was just released 🎉

**Comment by Elabajaba** [comment]
I can't reproduce the windows CI failure for the minimising example locally both on a dedicated GPU and forcing it to use WARP.

**Review by mockersf — APPROVED** [review]

**Comment by hymm** [comment]
can't reproduce on my laptop
```text
AdapterInfo { name: "NVIDIA GeForce RTX 3070 Ti Laptop GPU", vendor: 4318, device: 9376, device_type: DiscreteGpu, driver: "", driver_info: "", backend: Dx12 }
```

**Comment by JMS55** [comment]
Can't reproduce either
```
2023-10-01T20:09:55.009018Z  INFO bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce RTX 3080", vendor: 4318, device: 8710, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "536.67", backend: Vulkan }
2023-10-01T20:09:55.953721Z  INFO bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Windows 11 Home", kernel: "22621", cpu: "AMD Ryzen 5 2600 Six-Core Processor", core_count: "6", memory: "15.9 GiB" }
```

**Review by JMS55 — APPROVED** [review]

**Comment by mockersf** [comment]
re triggered CI, it's still failing with the same error:
```
2023-10-01T20:33:12.009607Z ERROR wgpu_hal::auxil::dxgi::exception: ID3D12CommandAllocator::Reset: A command allocator 0x000002ACC79B7860:'Unnamed ID3D12CommandAllocator Object' is being reset before previous executions associated with the allocator have completed. [ EXECUTION ERROR #552: COMMAND_ALLOCATOR_SYNC]    
2023-10-01T20:33:12.009841Z ERROR wgpu_hal::auxil::dxgi::exception: ID3D12Resource2::<final-release>: CORRUPTION: An ID3D12Resource object (0x000002ACC607F6E0:'(wgpu internal) Staging') is referenced by GPU operations in-flight on Command Queue (0x000002ACA6BA6CD0:'Unnamed ID3D12CommandQueue Object').  It is not safe to final-release objects that may have GPU operations pending.  This can result in application instability. [ EXECUTION ERROR #921: OBJECT_DELETED_WHILE_STILL_IN_USE]  
```

**Comment by Azorlogh** [comment]
Just to let you know, wgpu-hal v0.17.1 is currently affected by [another resize-related issue](https://github.com/gfx-rs/wgpu/issues/4180) which prevents on X11 + Vulkan.
It will be hopefully resolved soon in 0.17.2 as the fix PR is already there though

**Comment by cwfitzgerald** [comment]
Fixed in [v0.17.2](https://github.com/gfx-rs/wgpu/releases/tag/v0.17.2)

**Comment by Elabajaba** [comment]
Oof, the resizing example also works fine for me locally on a my gpu and on WARP. (`$env:WGPU_ADAPTER_NAME="Microsoft Basic Render Driver"`, and changing https://github.com/bevyengine/bevy/blob/f9e50e767b76382527bff93385247829461758a0/crates/bevy_render/src/renderer/mod.rs#L134 to use `wgpu::util::initialize_adapter_from_env` to test out warp locally)

**Review by superdump — APPROVED** [review]

File: crates/bevy_render/Cargo.toml
Comment (tbillington) [inline crates/bevy_render/Cargo.toml:63]: Should this use 17.2 per @cwfitzgerald comment?

Comment (Elabajaba) [reply crates/bevy_render/Cargo.toml:63]: 0.17.2 was `wgpu-hal` only, there's no 0.17.2 for `wgpu` (and cargo will automatically choose the highest patch version here).

Comment (tbillington) [reply crates/bevy_render/Cargo.toml:63]: Ah, my mistake, cheers

**Pull request merged**
